### DISCUSSION TITLE
Issues with pip install quarto-cli

### DISCUSSION THREAD

**Discussion by leemhmark** [quarto-dev/quarto-cli#11380] [Q&A] [answered]
### Description

I keep encountering the following build errors when trying to pip install quarto-cli on my Windows machine and I'm not sure how to resolve it, any suggestions would be great.
![quarto-cli](https://github.com/user-attachments/assets/9f0c6e51-c5ac-40d7-9193-b8ab4a1de826)

**Comment by mcanouil** [comment]
Unfortunately, a screenshot of an error is not enough.
Please fully describe your setup and the steps that lead to this error.

**Reply by leemhmark** [reply]
Ok, I was trying to install it within a conda environment:

conda create -n quarto-cli python=3.10
pip install quarto-cli

I'm currently using conda 22.9.0

**Reply by cderv** [reply]
If you are using conda there is a conda-forge version : https://anaconda.org/conda-forge/quarto

It is probably better to install in conda environment using conda than pip... 

I don't know if building wheel in conda context is a problem... 🤔 I'll try on my side once I am setup

**Reply by leemhmark** [reply]
Ok thanks for bringing that up, I'll give that a try.

Strange thing is it didn't work for me when I try to pip install outside a conda environment either (same error):
- opened cmd
- ran pip install quarto-cli

I'm using pip 22.2.2 (python 3.9)

**Reply by cderv** [reply]
I am on python 3.13 FWIW

**Reply by leemhmark** [reply]
I'm unable to use pip install either on python 3.13:

```
conda create -n quarto-py313 python=3.13
conda activate quarto-py313 python=3.13
pip install quarto-cli
```
Gave me the same error. I was however able to get it using conda install quarto. It works fine except that in VS Code, the shortcuts to render quarto documents don't work since they still want to pip install it, so I have to run the quarto commands manually on the terminal. So some minor inconveniences but still would like to resolve the issues with pip install.

**Reply by cderv** [reply]
> Gave me the same error. 

It seems like something in the conda environment prevent installing this type of package somehow. 🤔 

> It works fine except that in VS Code, the shortcuts to render quarto documents don't work since they still want to pip install it, so I have to run the quarto commands manually on the terminal.

There may be an issue in conda environment detection then. Do you have it correctly activated in VS CODE ? You have also a quarto extension setting that should allow you to configure the quarto binary to use for your project, as a project setting

**Reply by leemhmark** [reply]
So this is the setting you're referring to?

![image](https://github.com/user-attachments/assets/2593112c-9095-4610-9511-adb9c707eef3)

In my conda environment directories, there are 2 quarto binary files under C:\users\[my name]\.conda\envs\quarto-cli\Library\bin, quarto.js and quarto.cmd. I've set the path to both but it's still not working (Ctrl + Shift + K still says no quarto installed)

**Reply by cderv** [reply]
> So this is the setting you're referring to?

Yes 

> In my conda environment directories, there are 2 quarto binary files under C:\users[my name].conda\envs\quarto-cli\Library\bin, quarto.js and quarto.cmd

Binary to use is `quarto.cmd` here. 

> I've set the path to both but it's still not working (Ctrl + Shift + K still says no quarto installed)

If the setting is set, I find this odd... though QUARTO VS CODE extension may have some priority rules, especially regarding virtualenvironment that may be activated. 

Check if you have some python environment loaded in VSCODE that should not be. 

I'll try to reproduce, but I would expect the setting to work. 

@mcanouil does it ring a bell the setting not behaving 100% as it should ?

**Reply by mcanouil** [reply]
I have some ideas but I need to check things first as I don't have Conda (for kind of obvious reasons😅)

**Reply by mcanouil** [reply]
If you have Docker and the Devcontainer VSCode extension, let's try to remove some environments components that might be specific:

In `.devcontainer/devcontainer.json`:

```json
{
	"name": "Ubuntu",
	"image": "mcr.microsoft.com/devcontainers/base:jammy",
	"features": {
		"ghcr.io/devcontainers/features/conda:1": {
			"version": "latest"
		},
		"ghcr.io/devcontainers/features/python:1": {
			"installTools": true,
			"enableShared": true,
			"version": "latest"
		}
	}
}
```

Start the Devcontainer, in the terminal:

```bash
conda create -n quarto-py313
conda init zsh
conda activate quarto-py313
pip install quarto-cli
```

If this works, it means the issue lies in your Conda installation/configuration.
If it does not work, then it means the issue is in your VSCode settings.

What you might want to check is the following setting: `python.defaultInterpreterPath`.
It should be `python` (it's very rare to use something else when working with environments).
You can "reset setting".

In case it was not clear, all the above worked for me, so I am 100 % confident the issue is not on Quarto side.

**Reply by leemhmark** [reply]
I used the base environment pointing to the python.exe in my conda install directory. For reference I've also tried to activate the environment that has quarto installed, and it automatically set some environment variables but Ctrl + Shift + K still gave me the error:

![image](https://github.com/user-attachments/assets/6f6aaf27-b82d-462a-a90e-b0c4fb8ce3fb)

As for mcanouil's suggestion unfortunately I don't have Docker installed and I'm not allowed to use it in my organization. However note that as I mentioned earlier I wasn't able to pip install it outside a conda environment either:

1. open windows CMD
2. pip install quarto-cli

**Reply by mcanouil** [reply]
You are mixing issues here. Your last screenshot is unrelated to the issue in this thread. If you can't install Quarto, of course the Quarto VSCode extension won't find it.

Your environment is the issue, but without the means to "copy" it, it's goign to be hard to figure out the issue.

What's your Z volume? If it's a network/cloud drive, please test everything on your local drives. This adds complexities.

Also, why are you setting the `DENO_` environment variables? Did you set other things like that?

**Reply by leemhmark** [reply]
I wasn't setting the DENO_ environment variables, it does that automatically when I activate the environment. As mentioned earlier I was able to conda install quarto in this environment, the problem right now is VS Code can't find that installation. 

The terminal was on Z drive since my repo was there, but all my environments were created under C:\users, you can see the same problem when I tried to do it in C:\

![image](https://github.com/user-attachments/assets/62cad019-a95a-4c01-b3c7-917431e12209)

**Reply by cderv** [reply]
> Also, why are you setting the DENO_ environment variables? Did you set other things like that?

@mcanouil this happens when running `quarto.cmd`. No need to assume user's is doing it - it is us. `quarto.cmd` is a script that will run on windows using BAT. 

It will be especially used by conda-forge of Quarto which don't build an `.exe`.

Now why this script is ran ? It seems the conda environment does run quarto but it is possible there is an issue while running Quarto and so why the VSCODE extension is saying "not found" because quarto install is not running. 

@leemhmark you should take a step back and check that outside of VSCODE you get a running quarto installation. 

- If you are using conda, create a new clean environment in terminal, install quarto from conda-forge and check that it is working. 

- If you are not using conda, use python directly with a venv (see example below https://github.com/quarto-dev/quarto-cli/discussions/11380#discussioncomment-11238755) and install with `pip install`. **though you need to be sure when doing that you are not using any conda environment, conda binary or else**. 


Mixing conda and other install is not easy to handle on windows, I had really bad experience with that. 

So I would try to have things working the way you want in Terminal only. Then when it is working try again VSCODE. 

Also, just a reminder that there is no need to install quarto using pip or conda. You can install by downloading the right installer from the website, and install on your machine for your user. Installing using pip is advanced usage when pinning a version for a specific project is desired for example. 

Hope it helps.

**Reply by mcanouil** [reply, marked as answer]
~~Conda is setting these, hum, I really don't understand the appeal of this tool.~~
It's actually Quarto apparently.

Anyhow, you basically put aside the method to install Quarto via PIP, right?
You've managed to install Quarto via Conda-Forge as you said.
Do you know where Quarto CLI is in this setup (the path)? (probably something like `conda/envs/quarto-py313/bin/quarto`)

Using the same Docker image as before leads to everything working as intended in the terminal because that's where the environment is activated.

```sh
conda create -n quarto-py313
conda init zsh
conda activate quarto-py313
conda install conda-forge::quarto
```

Does `quarto --version` work in your activated environment?

<img width="1624" alt="image" src="https://github.com/user-attachments/assets/67b971d7-b3ae-42d9-a71b-b689ce60af17">

If you open a new terminal you'll see that the environment is no longer activated.

Now you need to figure out how to tell VSCode to use this environment everywhere.
VSCode does this natively for local environment when it detects them (`.conda` inside the workspace), but cannot do that when it's elsewhere, but still the tools won't be accessible to VSCode because it was started before the environment was activated.
In the end, this is still not a Quarto CLI issue.

See <https://quarto.org/docs/projects/virtual-environments.html#vs-code>.
This behaviour about Conda and VSCode (and extension setting not working) is documented.

**Reply by leemhmark** [reply]
Ok I'm able to make it work now, basically need to activate the conda environment with the quarto installed and then launch VS Code from there:

1. Open anaconda prompt
2. conda activate quarto-cli_py310
3. code .

Now VS Code launches and the shortcuts do work as it recognizes the Quarto install:
![image](https://github.com/user-attachments/assets/2ee8b5db-d2ea-48af-9786-2d3a9f921781)

This seems to be the only way so far, consider my issue resolved.

**Reply by mcanouil** [reply]
Yes, I think there is an issue open in the Quarto VSCode extension repository about opening VSCode from the command line where the environment has been activated.

Glad you found a way to make it work!

**Comment by cderv** [comment]
Right now I can't reproduce 

What I did on my windows environment using **uv** (https://docs.astral.sh/uv) for managing venv

````
cd $env:TEMP
mkdir test-quarto
cd test-quarto
uv venv
uv pip install quarto-cli
````

It installed latest quarto correctly. 

Verbose mode, also showed by not issue at the line where you got the error
````
DEBUG adding 'quarto_cli/share/deno_std/cache/deps/https/cdn.skypack.dev/010ab5a2e2b8bd9419eceeb3654e9d2f1412afd9065e89a8d3d03c58f29970dd'
DEBUG adding 'quarto_cli/share/deno_std/cache/deps/https/cdn.skypack.dev/010ab5a2e2b8bd9419eceeb3654e9d2f1412afd9065e89a8d3d03c58f29970dd.metadata.json'
DEBUG adding 'quarto_cli/share/deno_std/cache/deps/https/cdn.skypack.dev/075584d2a217fa87db56cb38b63393804f62af3308285711316daab6892cebb6'
DEBUG adding 'quarto_cli/share/deno_std/cache/deps/https/cdn.skypack.dev/075584d2a217fa87db56cb38b63393804f62af3308285711316daab6892cebb6.metadata.json'
````

So probably something related to your environment somehow.
